srml/balances/src/lib.rs

```diff
@@ -437,28 +489,50 @@ where
 		<ReservedBalance<T, I>>::get(who)
 	}
 
-	fn slash(who: &T::AccountId, value: Self::Balance) -> Option<Self::Balance> {
+	fn slash<S: OnUnbalancedDecrease<Self::Balance>>(
+		who: &T::AccountId,
+		value: Self::Balance
+	) -> Option<Self::Balance> {
 		let free_balance = Self::free_balance(who);
 		let free_slash = cmp::min(free_balance, value);
 		Self::set_free_balance(who, free_balance - free_slash);
-		Self::decrease_total_stake_by(free_slash);
-		if free_slash < value {
-			Self::slash_reserved(who, value - free_slash)
+		let res = if free_slash < value {
+			let remaining_slash = value - free_slash;
+			let reserved_balance = Self::reserved_balance(who);
+			let reserved_slash = cmp::min(reserved_balance, remaining_slash);
+			Self::set_reserved_balance(who, reserved_balance - reserved_slash);
+			if reserved_slash < remaining_slash {
+				Some(remaining_slash - reserved_slash)
+			} else {
+				None
+			}
 		} else {
 			None
-		}
+		};
+
+		// underflow should never happen, but it if does, there's nothing to be done here.
+		let _ = S::on_unbalanced_decrease(value - res.unwrap_or_default());
+		res
 	}
 
-	fn reward(who: &T::AccountId, value: Self::Balance) -> result::Result<(), &'static str> {
+	fn reward<S: OnUnbalancedIncrease<Self::Balance>>(
+		who: &T::AccountId,
+		value: Self::Balance
+	) -> result::Result<(), &'static str> {
 		if Self::total_balance(who).is_zero() {
 			return Err("beneficiary account must pre-exist");
 		}
+		S::on_unbalanced_increase(value)?;
 		Self::set_free_balance(who, Self::free_balance(who) + value);
-		Self::increase_total_stake_by(value);
 		Ok(())
 	}
 
-	fn increase_free_balance_creating(who: &T::AccountId, value: Self::Balance) -> UpdateBalanceOutcome {
+	fn increase_free_balance_creating<S: OnUnbalancedIncrease<Self::Balance>>(
+		who: &T::AccountId,
+		value: Self::Balance
+	) -> UpdateBalanceOutcome {
+		// Should be infallible if our state isn't corruption. Not much we can do if state it corrupted.
+		let _ = S::on_unbalanced_increase(value);
```

Comment:
OK I was wrong for this method:
the doc said: `NOTE: This assumes that the total stake remains unchanged after this operation`
because when I decoupled treasury it was not increase stake
https://github.com/paritytech/substrate/blob/e01b4501127fcfc8c62ae18fea8a071c8cf18c34/srml/treasury/src/lib.rs#L226-L232

But finally it seems what I thought as intentionnal was a bug.
So this method shouldn't exist anymore and be renamed to reward_creating. (or reward_create_account as suggested by @joepetrowski )

Also this confirm the point of @shawntabrizi as TotalIssuance, ReservedBalance and FreeBalance we could make ourself safer even by providing method that will operate them on them coherently. We can also provides bypass method for some optimisation (but most of usages can use safe abstraction actually as we almost always call modification of total issuance and then modification of free balance) and we can even name those bypass method with a unguaranted_ or other very warnant prefix.

This would make reviewing easier

Reply:
hmm safe operator could be implemented in another PR obviously but renaming of increase_free_balance_creating to reward_create_account seems relevant to me

Reply:
They are not dependent storages. Total issuance includes balances outside of the basic AccountId mapping such as that in parachains.

Reply:
In any case, this is outside the scope of this PR.